chains/evm/message/mayan.go

```diff
@@ -0,0 +1,328 @@
+	"github.com/sprintertech/sprinter-signing/tss/ecdsa/signing"
+	"github.com/sygmaprotocol/sygma-core/relayer/message"
+	"github.com/sygmaprotocol/sygma-core/relayer/proposal"
+)
+
+var (
+	BPS_DENOMINATOR = big.NewInt(10000)
+)
+
+type MayanContract interface {
+	DecodeFulfillCall(calldata []byte) (*contracts.MayanFulfillParams, *contracts.MayanFulfillMsg, error)
+	GetOrder(
+		msg *contracts.MayanFulfillMsg,
+		swap *mayan.MayanSwap,
+		srcTokenDecimals uint8,
+	) (*contracts.MayanOrder, error)
+}
+
+type SwapFetcher interface {
+	GetSwap(hash string) (*mayan.MayanSwap, error)
+}
+
+type MayanMessageHandler struct {
+	client  EventFilterer
+	chainID uint64
+
+	mayanPools          map[uint64]common.Address
+	liqudityPools       map[uint64]common.Address
+	confirmationWatcher ConfirmationWatcher
+	tokenStore          config.TokenStore
+	mayanDecoder        MayanContract
+	swapFetcher         SwapFetcher
+
+	coordinator Coordinator
+	host        host.Host
+	comm        comm.Communication
+	fetcher     signing.SaveDataFetcher
+
+	sigChn chan any
+}
+
+func NewMayanMessageHandler(
+	chainID uint64,
+	client EventFilterer,
+	liqudityPools map[uint64]common.Address,
+	mayanPools map[uint64]common.Address,
+	coordinator Coordinator,
+	host host.Host,
+	comm comm.Communication,
+	fetcher signing.SaveDataFetcher,
+	confirmationWatcher ConfirmationWatcher,
+	tokenStore config.TokenStore,
+	mayanDecoder MayanContract,
+	swapFetcher SwapFetcher,
+	sigChn chan any,
+) *MayanMessageHandler {
+	return &MayanMessageHandler{
+		chainID:             chainID,
+		client:              client,
+		mayanPools:          mayanPools,
+		liqudityPools:       liqudityPools,
+		coordinator:         coordinator,
+		host:                host,
+		comm:                comm,
+		fetcher:             fetcher,
+		sigChn:              sigChn,
+		confirmationWatcher: confirmationWatcher,
+		mayanDecoder:        mayanDecoder,
+		swapFetcher:         swapFetcher,
+		tokenStore:          tokenStore,
+	}
+}
+
+// HandleMessage finds the Mayan deposit with the according deposit ID and starts
+// the MPC signature process for it. The result will be saved into the signature
+// cache through the result channel.
+func (h *MayanMessageHandler) HandleMessage(m *message.Message) (*proposal.Proposal, error) {
+	data := m.Data.(*MayanData)
+	txHash := common.HexToHash(data.DepositTxHash)
+
+	err := h.notify(data)
+	if err != nil {
+		log.Warn().Msgf("Failed to notify relayers because of %s", err)
+	}
+
+	calldataBytes, err := hex.DecodeString(data.Calldata)
+	if err != nil {
+		data.ErrChn <- err
+		return nil, err
+	}
+
+	params, msg, err := h.mayanDecoder.DecodeFulfillCall(calldataBytes)
+	if err != nil {
+		data.ErrChn <- err
+		return nil, err
+	}
+	swap, err := h.swapFetcher.GetSwap(txHash.Hex())
+	if err != nil {
+		data.ErrChn <- err
+		return nil, err
+	}
+
+	symbol, token, err := h.tokenStore.ConfigByAddress(h.chainID, common.BytesToAddress(msg.TokenIn[12:]))
```

Comment:
maybe extract into variable
`tokenInAddress := common.BytesToAddress(msg.TokenIn[12:])`
as you use it in one more place down